File: zeroinstall/zerostore/optimise.py
"""Optimise the cache."""

# Copyright (C) 2009, Thomas Leonard
# See the README file for details, or visit http://0install.net.

from __future__ import print_function

from zeroinstall import _, logger
import os, sys

def _already_linked(a, b):
	"""@type a: str
	@type b: str
	@rtype: bool"""
	ai = os.stat(a)
	bi = os.stat(b)
	return (ai.st_dev, ai.st_ino) == (bi.st_dev, bi.st_ino)

def _byte_identical(a, b):
	"""@type a: str
	@type b: str
	@rtype: bool"""
	with open(a, 'rb') as af:
		with open(b, 'rb') as bf:
			while True:
				adata = af.read(100)
				bdata = bf.read(100)
				if adata != bdata:
					return False
				if not adata:
					return True

def _link(a, b, tmpfile):
	"""Keep 'a', delete 'b' and hard-link to 'a'
	@type a: str
	@type b: str
	@type tmpfile: str"""
	if not _byte_identical(a, b):
		logger.warning(_("Files should be identical, but they're not!\n%(file_a)s\n%(file_b)s"), {'file_a': a, 'file_b': b})

	b_dir = os.path.dirname(b)
	old_mode = os.lstat(b_dir).st_mode
	os.chmod(b_dir, old_mode | 0o200)	# Need write access briefly
	try:
		os.link(a, tmpfile)
		try:
			os.rename(tmpfile, b)
		except:
			os.unlink(tmpfile)
			raise
	finally:
		os.chmod(b_dir, old_mode)

def optimise(impl_dir):
	"""Scan an implementation cache directory for duplicate files, and
	hard-link any duplicates together to save space.
	@param impl_dir: a $cache/0install.net/implementations directory
	@type impl_dir: str
	@return: (unique bytes, duplicated bytes, already linked, manifest size)
	@rtype: (int, int, int, int)"""

	first_copy = {}		# TypeDigest -> Path
	dup_size = uniq_size = already_linked = man_size = 0

	import random
	from zeroinstall.zerostore import BadDigest, parse_algorithm_digest_pair

	for x in range(10):
		tmpfile = os.path.join(impl_dir, 'optimise-%d' % random.randint(0, 1000000))
		if not os.path.exists(tmpfile):
			break
	else:
		raise Exception(_("Can't generate unused tempfile name!"))

	dirs = os.listdir(impl_dir)
	total = len(dirs)
	msg = ""
	def clear():
		print("\r" + (" " * len(msg)) + "\r", end='')
	for i, impl in enumerate(dirs):
		clear()
		msg = _("[%(done)d / %(total)d] Reading manifests...") % {'done': i, 'total': total}
		print(msg, end='')
		sys.stdout.flush()

		try:
			alg, manifest_digest = parse_algorithm_digest_pair(impl)
		except BadDigest:
			logger.warning(_("Skipping non-implementation '%s'"), impl)
			continue
		manifest_path = os.path.join(impl_dir, impl, '.manifest')
		try:
			ms = open(manifest_path, 'rt')
		except OSError as ex:
			logger.warning(_("Failed to read manifest file '%(manifest_path)s': %(exception)s"), {'manifest': manifest_path, 'exception': str(ex)})
			continue

		if alg == 'sha1':
			ms.close()
			continue

		man_size += os.path.getsize(manifest_path)

		dir = ""
		for line in ms:
			if line[0] == 'D':
				itype, path = line.split(' ', 1)
				assert path.startswith('/')
				dir = path[1:-1]	# Strip slash and newline
				continue

			if line[0] == "S":
				itype, digest, size, rest = line.split(' ', 3)
				uniq_size += int(size)
				continue

			assert line[0] in "FX"

			itype, digest, mtime, size, path = line.split(' ', 4)
			path = path[:-1]	# Strip newline
			size = int(size)

			key = (itype, digest, mtime, size)
			loc_path = (impl, dir, path)

			first_loc = first_copy.get(key, None)
			if first_loc:
				first_full = os.path.join(impl_dir, *first_loc)
				new_full = os.path.join(impl_dir, *loc_path)
				if _already_linked(first_full, new_full):
					already_linked += size
				else:
					_link(first_full, new_full, tmpfile)
					dup_size += size
			else:
				first_copy[key] = loc_path
				uniq_size += size

		ms.close()
	clear()
	return (uniq_size, dup_size, already_linked, man_size)

File: src/cli/optimise.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

(** The "0install store optimise" command *)

open Options
open Zeroinstall.General
open Support
open Support.Common

module U = Support.Utils
module Manifest = Zeroinstall.Manifest

let are_already_linked (system:system) a b =
  match system#lstat a, system#lstat b with
  | Some ai, Some bi ->
      ai.Unix.st_dev = bi.Unix.st_dev &&
      ai.Unix.st_ino = bi.Unix.st_ino
  | _ -> Safe_exn.failf "Missing files comparing %s and %s" a b

let are_byte_identical (system:system) a b =
  a |> system#with_open_in [Open_rdonly;Open_binary] (fun ia ->
    b |> system#with_open_in [Open_rdonly;Open_binary] (fun ib ->
      let rec loop () =
        let adata = U.read_upto 100 ia in
        let bdata = U.read_upto 100 ib in
        if adata <> bdata then false
        else if adata = "" then true
        else loop () in
      loop ()
    )
  )

(** Keep 'a', delete 'b' and hard-link to 'a' *)
let link (system:system) a b ~tmpfile =
  if not (are_byte_identical system a b) then (
    log_warning "Files should be identical, but they're not!\n%s\n%s" a b
  ) else (
    let b_dir = Filename.dirname b in
    let b_dir_info = system#lstat b_dir |? lazy (Safe_exn.failf "Can't lstat %s!" b_dir) in
    let old_mode = b_dir_info.Unix.st_perm in
    system#chmod b_dir (old_mode lor 0o200);	(* Need write access briefly *)
    U.finally_do (fun () -> system#chmod b_dir old_mode) ()
      (fun () ->
        system#hardlink a tmpfile;
        if on_windows then system#unlink b;
        try system#rename tmpfile b
        with ex ->
          system#unlink tmpfile;
          raise ex
      )
  )

type size = Int64.t

type report = {
  mutable uniq_size : size;
  mutable dup_size : size;
  mutable already_linked : size;
  mutable man_size : size;
}

let optimise_impl system stats first_copy ~tmpfile impl_dir impl =
  let digest = Manifest.parse_digest impl in
  let manifest_path = impl_dir +/ impl +/ ".manifest" in
  let ms = U.read_file system manifest_path in

  if fst digest = "sha1" then ()
  else (
    stats.man_size <- Int64.add stats.man_size (Int64.of_int @@ String.length ms);

    let rec scan dir_path dir_items =
      dir_items |> List.iter (fun (name, item) ->
        match item with
        | `Dir items -> scan (dir_path +/ name) items
        | `Symlink (_, size) -> stats.uniq_size <- Int64.add stats.uniq_size size
        | `File info ->
            let (_x, _hash, _mtime, size) = info in
            let new_full = dir_path +/ name in
            match Hashtbl.find_opt first_copy info with
            | None ->
                Hashtbl.add first_copy info new_full;
                stats.uniq_size <- Int64.add stats.uniq_size size
            | Some first_full ->
                if are_already_linked system first_full new_full then (
                  stats.already_linked <- Int64.add stats.already_linked size
                ) else (
                  link system first_full new_full ~tmpfile;
                  stats.dup_size <- Int64.add stats.dup_size size
                )
      ) in
    scan (impl_dir +/ impl) (Manifest.parse_manifest ms)
  )

(** Scan an implementation cache directory for duplicate files, and
    hard-link any duplicates together to save space. *)
let optimise system impl_dir =
  let first_copy = Hashtbl.create 1024 in		(* `File tuple -> path *)
  let stats = {
    uniq_size = Int64.zero;
    dup_size = Int64.zero;
    already_linked = Int64.zero;
    man_size = Int64.zero;
  } in

  (* Find an unused filename we can used during linking *)
  let tmpfile =
    let rec mktmp = function
      | 0 -> Safe_exn.failf "Failed to generate temporary file name!"
      | n ->
          let tmppath = impl_dir +/ Printf.sprintf "optimise-%x" (Random.int 0x3fffffff) in
          if system#file_exists tmppath then mktmp (n - 1)
          else tmppath in
    mktmp 10 in

  let dirs =
    match system#readdir impl_dir with
    | Error ex -> raise ex
    | Ok items -> items in
  let total = Array.length dirs in
  let msg = ref "" in
  let clear () =
    let blank = String.make (String.length !msg) ' ' in
    Printf.printf "\r%s\r%!" blank
  in
  dirs |> Array.iteri (fun i impl ->
    clear ();
    Printf.printf "[%d / %d] Reading manifests...%!" i total;
    try optimise_impl system stats first_copy ~tmpfile impl_dir impl
    with Safe_exn.T e -> clear (); log_warning "Skipping '%s': %a" impl Safe_exn.pp e
  );
  clear ();
  stats

let handle options flags args =
  Support.Argparse.iter_options flags (Common_options.process_common_option options);
  let config = options.config in
  let system = config.system in

  let cache_dir =
    match args with
    | [] -> List.hd options.config.stores
    | [dir] -> U.abspath system dir
    | _ -> raise (Support.Argparse.Usage_error 1) in

  if not (U.is_dir system cache_dir) then (
    Safe_exn.failf "Not a directory: '%s'" cache_dir
  );

  let impl_name = Filename.basename cache_dir in
  if impl_name <> "implementations" then (
    Safe_exn.failf "Cache directory should be named 'implementations', not '%s' (in '%s')" impl_name cache_dir
  );

  let print fmt = Format.fprintf options.stdout (fmt ^^ "@.") in
  print "Optimising %s" cache_dir;

  let {uniq_size; dup_size; already_linked; man_size} = optimise config.system cache_dir in
  print "Original size  : %s (excluding the %s of manifests)" (U.format_size (Int64.add uniq_size dup_size)) (U.format_size man_size);
  print "Already saved  : %s" (U.format_size already_linked);
  if dup_size = Int64.zero then (
    print "No duplicates found; no changes made."
  ) else (
    print "Optimised size : %s" (U.format_size uniq_size);
    let perc = (100.0 *. Int64.to_float dup_size) /. (Int64.to_float uniq_size +. Int64.to_float dup_size) in
    print "Space freed up : %s (%.2f%%)" (U.format_size dup_size) perc
  );
  print "Optimisation complete."
